File: src/pi_wave/rpc.py
"""Async client for pi's RPC mode (JSONL over stdin/stdout).

Framing per pi docs/rpc.md: records are separated by LF (\n) only. asyncio's
stream readline() splits on \n, which is protocol-compliant.

One PiRpcSession = one assignment agent. The session is short-lived per
assignment (and per fix round we re-prompt the same process, mirroring
delegate-wave's "one consolidated fix prompt to the originating agent").
"""

from __future__ import annotations

import asyncio
import json
import os
from pathlib import Path
from typing import Any, Callable

EventHook = Callable[[str, dict[str, Any]], None]

# pi's MCP support (--mcp-config) is registered by the pi-mcp-adapter package.
# With --no-extensions, package discovery is off but explicit -e paths still
# load — so we load ONLY the adapter (never dispatch-wave: recursion guard
# stays intact) and hand it the assignment's MCP config.
MCP_ADAPTER_ENTRY = os.environ.get(
    "PI_WAVE_MCP_ADAPTER",
    "~/.pi/agent/npm/node_modules/pi-mcp-adapter/index.ts",
)


def mcp_flags(mcp_config: str) -> list[str]:
    if not mcp_config:
        return []
    entry = Path(MCP_ADAPTER_ENTRY).expanduser()
    if not entry.is_file():
        raise PiRpcError(
            f"mcp_config requires the pi-mcp-adapter package but its entry was not "
            f"found at {entry} (install with: pi install npm:pi-mcp-adapter, or set PI_WAVE_MCP_ADAPTER)"
        )
    return ["--no-extensions", "-e", str(entry), "--mcp-config", str(Path(mcp_config).expanduser())]


class PiRpcError(RuntimeError):
    pass


class PiRpcSession:
    def __init__(
        self,
        *,
        agent_name: str,
        cwd: str,
        model: str,
        provider: str | None = None,
        thinking: str = "high",
        load_extensions: bool = False,
        mcp_config: str = "",
        on_event: EventHook | None = None,
    ) -> None:
        self.agent_name = agent_name
        self._cwd = cwd
        self._model = model
        self._provider = provider
        self._thinking = thinking
        self._load_extensions = load_extensions
        self._mcp_config = mcp_config
        self._on_event = on_event

        self.proc: asyncio.subprocess.Process | None = None
        self._next_id = 0
        self._waiters: dict[str, asyncio.Future[dict[str, Any]]] = {}
        self._events: asyncio.Queue[dict[str, Any]] = asyncio.Queue()
        self._stderr_tail: list[str] = []
        self._last_text: str | None = None
        self._turns = 0
        self._closed = False

    # -- lifecycle ----------------------------------------------------------

    async def start(self) -> dict[str, Any]:
        argv = [
            "pi", "--mode", "rpc", "--no-session",
            "--model", self._model,
            "--thinking", self._thinking,
        ]
        if self._provider:
            argv += ["--provider", self._provider]
        # Recursion guard: assignment agents must never load dispatch_wave
        # (soft rule in prompts, hard guarantee here).
        if not self._load_extensions:
            argv.append("--no-extensions")
        argv += mcp_flags(self._mcp_config)

        self.proc = await asyncio.create_subprocess_exec(
            *argv,
            cwd=self._cwd,
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
            limit=2**22,
        )
        asyncio.get_running_loop().create_task(self._drain_stderr())
        asyncio.get_running_loop().create_task(self._read_stdout())
        state = await self.request({"type": "get_state"}, timeout=30)
        return state["data"]

    def verify_model(self, state: dict[str, Any]) -> None:
        """Fail fast when pi did not resolve the requested model (the RPC
        analogue of delegate-wave's 'check the pane footer' rule)."""
        got = (state.get("model") or {}).get("id")
        want = self._model.split("/")[-1] if "/" in self._model else self._model
        if got != want:
            raise PiRpcError(
                f"[{self.agent_name}] model mismatch: requested '{self._model}' "
                f"but pi resolved '{got}' — fix the plan's model id "
                f"(check with: pi --list-models {want})"
            )

    async def close(self) -> None:
        if self._closed or self.proc is None:
            return
        self._closed = True
        try:
            await asyncio.wait_for(self.request({"type": "abort"}, timeout=5), timeout=3)
        except Exception:
            pass
        try:
            self.proc.terminate()
            await asyncio.wait_for(self.proc.wait(), timeout=5)
        except Exception:
            try:
                self.proc.kill()
            except Exception:
                pass

    # -- requests / events --------------------------------------------------

    async def request(self, cmd: dict[str, Any], timeout: float = 120.0) -> dict[str, Any]:
        if self.proc is None or self.proc.stdin is None:
            raise PiRpcError(f"[{self.agent_name}] session not started")
        self._next_id += 1
        rid = f"r{self._next_id}"
        fut: asyncio.Future[dict[str, Any]] = asyncio.get_running_loop().create_future()
        self._waiters[rid] = fut
        self.proc.stdin.write((json.dumps({**cmd, "id": rid}) + "\n").encode())
        await self.proc.stdin.drain()
        try:
            resp = await asyncio.wait_for(fut, timeout)
        finally:
            self._waiters.pop(rid, None)
        if resp.get("success") is False:
            raise PiRpcError(f"[{self.agent_name}] {cmd.get('type')} failed: {resp.get('error')}")
        return resp

    async def prompt_and_settle(self, message: str, timeout: float) -> None:
        """Send one prompt, then block until agent_settled (no retry,
        compaction retry, or queued continuation remains)."""
        if self.proc is None:
            raise PiRpcError(f"[{self.agent_name}] session not started")
        await self.request({"type": "prompt", "message": message}, timeout=30)
        loop = asyncio.get_running_loop()
        deadline = loop.time() + timeout
        while True:
            remaining = deadline - loop.time()
            if remaining <= 0:
                await self._abort_and_drain()
                raise TimeoutError(
                    f"[{self.agent_name}] prompt did not settle within {timeout:.0f}s"
                )
            ev = await self._next_event(remaining)
            self._absorb_event(ev)
            if ev.get("type") == "agent_settled":
                return

    async def last_text(self) -> str:
        if self._last_text is None:
            resp = await self.request({"type": "get_last_assistant_text"}, timeout=30)
            self._last_text = resp["data"].get("text") or ""
        return self._last_text

    async def stats(self) -> dict[str, Any]:
        try:
            resp = await self.request({"type": "get_session_stats"}, timeout=30)
            return resp["data"]
        except Exception:
            return {}

    # -- internals ------------------------------------------------------------

    def _absorb_event(self, ev: dict[str, Any]) -> None:
        etype = ev.get("type")
        if etype == "message_end":
            msg = ev.get("message") or {}
            if msg.get("role") == "assistant":
                text = "".join(
                    c.get("text", "") for c in msg.get("content", []) if c.get("type") == "text"
                )
                self._last_text = text
        elif etype == "turn_end":
            self._turns += 1
            if self._on_event:
                self._on_event(self.agent_name, {"type": "progress", "turn": self._turns})

    async def _abort_and_drain(self) -> None:
        try:
            await self.request({"type": "abort"}, timeout=5)
            loop = asyncio.get_running_loop()
            deadline = loop.time() + 15
            while loop.time() < deadline:
                ev = await self._next_event(deadline - loop.time())
                if ev.get("type") == "agent_settled":
                    return
        except Exception:
            pass

    async def _next_event(self, timeout: float) -> dict[str, Any]:
        try:
            ev = await asyncio.wait_for(self._events.get(), timeout)
        except asyncio.TimeoutError:
            raise TimeoutError(f"[{self.agent_name}] no event within {timeout:.1f}s") from None
        if ev.get("type") == "__process_exit__":
            raise PiRpcError(
                f"[{self.agent_name}] pi process exited unexpectedly; "
                f"stderr tail: {''.join(self._stderr_tail)[-800:]}"
            )
        return ev

    async def _read_stdout(self) -> None:
        assert self.proc is not None and self.proc.stdout is not None
        try:
            while True:
                line = await self.proc.stdout.readline()
                if not line:  # EOF
                    break
                text = line.decode("utf-8", errors="replace").rstrip("\n")
                if text.endswith("\r"):
                    text = text[:-1]
                if not text.strip():
                    continue
                try:
                    obj = json.loads(text)
                except json.JSONDecodeError:
                    continue
                if obj.get("type") == "response" and "id" in obj:
                    fut = self._waiters.pop(str(obj["id"]), None)
                    if fut and not fut.done():
                        fut.set_result(obj)
                else:
                    await self._events.put(obj)
        finally:
            for fut in self._waiters.values():
                if not fut.done():
                    fut.set_exception(PiRpcError(f"[{self.agent_name}] session ended before response"))
            await self._events.put({"type": "__process_exit__"})

    async def _drain_stderr(self) -> None:
        assert self.proc is not None and self.proc.stderr is not None
        while True:
            line = await self.proc.stderr.readline()
            if not line:
                return
            self._stderr_tail.append(line.decode("utf-8", errors="replace"))
            if len(self._stderr_tail) > 200:
                self._stderr_tail.pop(0)

File: engine/rpc.ts
// Async client for pi's RPC mode (JSONL over stdin/stdout).
//
// Framing per pi docs/rpc.md: records are separated by LF (\n) only, so we
// split on \n ourselves (node:readline would also split on \r and U+2028).
//
// One PiRpcSession = one assignment agent. The session is short-lived per
// assignment (and per fix round we re-prompt the same process, mirroring
// delegate-wave's "one consolidated fix prompt to the originating agent").

import { type ChildProcessWithoutNullStreams, spawn } from "node:child_process";
import { statSync } from "node:fs";
import os from "node:os";
import path from "node:path";
import { AgentTimeoutError } from "./errors.ts";

type Json = Record<string, any>;
export type EventHook = (agentName: string, ev: Json) => void;

const expandUser = (p: string) =>
  p === "~" || p.startsWith("~/") ? path.join(os.homedir(), p.slice(1)) : p;

// pi's MCP support (--mcp-config) is registered by the pi-mcp-adapter package.
// With --no-extensions, package discovery is off but explicit -e paths still
// load - so we load ONLY the adapter (never dispatch-wave: recursion guard
// stays intact) and hand it the assignment's MCP config.
export const MCP_ADAPTER_ENTRY =
  process.env.PI_WAVE_MCP_ADAPTER ?? "~/.pi/agent/npm/node_modules/pi-mcp-adapter/index.ts";

export class PiRpcError extends Error {
  override name = "PiRpcError";
}

export function mcpFlags(mcpConfig: string): string[] {
  if (!mcpConfig) return [];
  const entry = expandUser(MCP_ADAPTER_ENTRY);
  if (!statSync(entry, { throwIfNoEntry: false })?.isFile()) {
    throw new PiRpcError(
      `mcp_config requires the pi-mcp-adapter package but its entry was not ` +
        `found at ${entry} (install with: pi install npm:pi-mcp-adapter, or set PI_WAVE_MCP_ADAPTER)`,
    );
  }
  return ["--no-extensions", "-e", entry, "--mcp-config", expandUser(mcpConfig)];
}

/** Every "provider/model" pi can run here, from `pi --list-models` (which
 * lists only providers with working auth), or null when the list cannot be
 * read - callers then skip the check rather than block the run. */
export function listPiModels(piBin = "pi", timeoutS = 30): Promise<Set<string> | null> {
  return new Promise((resolve) => {
    const proc = spawn(piBin, ["--list-models"], { stdio: ["ignore", "pipe", "ignore"] });
    let out = "";
    proc.stdout.setEncoding("utf8").on("data", (c: string) => (out += c));
    const timer = setTimeout(() => proc.kill("SIGKILL"), timeoutS * 1000);
    proc.on("error", () => {
      clearTimeout(timer);
      resolve(null);
    });
    proc.on("close", (code) => {
      clearTimeout(timer);
      const [header, ...rows] = out.trim().split("\n");
      if (code !== 0 || !header?.trim().startsWith("provider")) return resolve(null);
      const models = new Set<string>();
      for (const row of rows) {
        const [provider, model] = row.trim().split(/\s+/);
        if (provider && model) models.add(`${provider}/${model}`);
      }
      resolve(models);
    });
  });
}

export interface PiRpcSessionOptions {
  agentName: string;
  cwd: string;
  model: string;
  provider?: string | null;
  thinking?: string;
  loadExtensions?: boolean;
  mcpConfig?: string;
  onEvent?: EventHook;
  /** Executable to spawn; tests point this at a fake pi. */
  piBin?: string;
}

const PROCESS_EXIT = "__process_exit__";

/** Unbounded FIFO of events with a timed async get(). */
class EventQueue {
  private items: Json[] = [];
  private waiter: ((ev: Json) => void) | null = null;

  put(ev: Json): void {
    if (this.waiter) {
      const w = this.waiter;
      this.waiter = null;
      w(ev);
    } else {
      this.items.push(ev);
    }
  }

  /** Resolves with the next event, or null after timeoutMs. */
  get(timeoutMs: number): Promise<Json | null> {
    const ev = this.items.shift();
    if (ev) return Promise.resolve(ev);
    return new Promise((resolve) => {
      const timer = setTimeout(() => {
        this.waiter = null;
        resolve(null);
      }, Math.max(0, timeoutMs));
      this.waiter = (e) => {
        clearTimeout(timer);
        resolve(e);
      };
    });
  }
}

/** Resolves with fn's result, or rejects with onTimeout() after ms. */
function withTimeout<T>(p: Promise<T>, ms: number, onTimeout: () => Error): Promise<T> {
  let timer: NodeJS.Timeout;
  const t = new Promise<never>((_, reject) => {
    timer = setTimeout(() => reject(onTimeout()), ms);
  });
  return Promise.race([p, t]).finally(() => clearTimeout(timer));
}

export class PiRpcSession {
  readonly agentName: string;
  proc: ChildProcessWithoutNullStreams | null = null;

  private readonly opts: PiRpcSessionOptions;
  private nextId = 0;
  private waiters = new Map<string, { resolve: (r: Json) => void; reject: (e: Error) => void }>();
  private events = new EventQueue();
  private stderrTail: string[] = [];
  private lastTextCache: string | null = null;
  private turns = 0;
  private closed = false;
  private ended = false;
  private exited: Promise<void> | null = null;

  constructor(opts: PiRpcSessionOptions) {
    this.opts = opts;
    this.agentName = opts.agentName;
  }

  // -- lifecycle ----------------------------------------------------------

  async start(): Promise<Json> {
    const { model, provider, thinking = "high", loadExtensions = false, mcpConfig = "" } = this.opts;
    const argv = ["--mode", "rpc", "--no-session", "--model", model, "--thinking", thinking];
    if (provider) argv.push("--provider", provider);
    // Recursion guard: assignment agents must never load dispatch_wave
    // (soft rule in prompts, hard guarantee here).
    if (!loadExtensions) argv.push("--no-extensions");
    argv.push(...mcpFlags(mcpConfig));

    const proc = spawn(this.opts.piBin ?? "pi", argv, { cwd: this.opts.cwd, stdio: "pipe" });
    // A stream 'error' with no listener is an uncaught exception, which
    // would take down the host process (pi, when run from the extension).
    // Write failures (EPIPE after pi died) already reach request() through
    // the write callback, and a dead pi surfaces through 'close'.
    for (const emitter of [proc, proc.stdin, proc.stdout, proc.stderr]) emitter.on("error", () => {});
    this.exited = new Promise((resolve) => proc.on("close", () => resolve()));
    this.drainStderr(proc);
    this.readStdout(proc);
    // Spawn failures (e.g. ENOENT) surface as an 'error' event, not a throw.
    await new Promise<void>((resolve, reject) => {
      proc.once("spawn", resolve);
      proc.once("error", (e) => reject(new PiRpcError(`[${this.agentName}] cannot start pi: ${e.message}`)));
    });
    this.proc = proc;
    const state = await this.request({ type: "get_state" }, 30);
    return state.data;
  }

  /** Fail fast when pi did not resolve the requested model (the RPC
   * analogue of delegate-wave's 'check the pane footer' rule). */
  verifyModel(state: Json): void {
    const got = state?.model?.id;
    const want = this.opts.model.split("/").at(-1);
    if (got !== want) {
      throw new PiRpcError(
        `[${this.agentName}] model mismatch: requested '${this.opts.model}' ` +
          `but pi resolved '${got}' - fix the plan's model id ` +
          `(check with: pi --list-models ${want})`,
      );
    }
  }

  async close(): Promise<void> {
    const proc = this.proc;
    if (this.closed || proc === null) return;
    this.closed = true;
    try {
      await withTimeout(this.request({ type: "abort" }, 5), 3_000, () => new Error("abort timeout"));
    } catch {
      // best effort: the process may already be gone
    }
    proc.kill("SIGTERM");
    try {
      await withTimeout(this.exited!, 5_000, () => new Error("exit timeout"));
    } catch {
      proc.kill("SIGKILL");
    }
  }

  // -- requests / events --------------------------------------------------

  async request(cmd: Json, timeoutS = 120): Promise<Json> {
    const proc = this.proc;
    if (proc === null) throw new PiRpcError(`[${this.agentName}] session not started`);
    if (this.ended) throw new PiRpcError(`[${this.agentName}] session ended before response`);
    const rid = `r${++this.nextId}`;
    const resp = new Promise<Json>((resolve, reject) => this.waiters.set(rid, { resolve, reject }));
    // If the write fails we never await resp; its later rejection must not
    // become an unhandled rejection.
    resp.catch(() => {});
    let result: Json;
    try {
      await new Promise<void>((resolve, reject) =>
        proc.stdin.write(JSON.stringify({ ...cmd, id: rid }) + "\n", (e) =>
          e ? reject(new PiRpcError(`[${this.agentName}] cannot send ${cmd.type}: ${e.message}`)) : resolve(),
        ),
      );
      result = await withTimeout(
        resp,
        timeoutS * 1000,
        () => new AgentTimeoutError(`[${this.agentName}] ${cmd.type} got no response within ${timeoutS}s`),
      );
    } finally {
      this.waiters.delete(rid);
    }
    if (result.success === false) {
      throw new PiRpcError(`[${this.agentName}] ${cmd.type} failed: ${result.error}`);
    }
    return result;
  }

  /** Send one prompt, then block until agent_settled (no retry,
   * compaction retry, or queued continuation remains). */
  async promptAndSettle(message: string, timeoutS: number): Promise<void> {
    if (this.proc === null) throw new PiRpcError(`[${this.agentName}] session not started`);
    await this.request({ type: "prompt", message }, 30);
    const deadline = performance.now() + timeoutS * 1000;
    while (true) {
      const remaining = deadline - performance.now();
      if (remaining <= 0) {
        await this.abortAndDrain();
        throw new AgentTimeoutError(
          `[${this.agentName}] prompt did not settle within ${Math.round(timeoutS)}s`,
        );
      }
      // A wait that runs out loops back to the deadline check above, so a
      // stuck agent is aborted before we give up on it.
      const ev = await this.nextEvent(remaining);
      if (ev === null) continue;
      this.absorbEvent(ev);
      if (ev.type === "agent_settled") return;
    }
  }

  async lastText(): Promise<string> {
    if (this.lastTextCache !== null) return this.lastTextCache;
    const resp = await this.request({ type: "get_last_assistant_text" }, 30);
    const text: string = resp.data?.text || "";
    this.lastTextCache = text;
    return text;
  }

  async stats(): Promise<Json> {
    try {
      const resp = await this.request({ type: "get_session_stats" }, 30);
      return resp.data;
    } catch {
      return {};
    }
  }

  // -- internals ------------------------------------------------------------

  private absorbEvent(ev: Json): void {
    if (ev.type === "message_end") {
      const msg = ev.message ?? {};
      if (msg.role === "assistant") {
        this.lastTextCache = (msg.content ?? [])
          .filter((c: Json) => c.type === "text")
          .map((c: Json) => c.text ?? "")
          .join("");
      }
    } else if (ev.type === "turn_end") {
      this.turns += 1;
      this.opts.onEvent?.(this.agentName, { type: "progress", turn: this.turns });
    }
  }

  private async abortAndDrain(): Promise<void> {
    try {
      await this.request({ type: "abort" }, 5);
      const deadline = performance.now() + 15_000;
      while (performance.now() < deadline) {
        const ev = await this.nextEvent(deadline - performance.now());
        if (ev === null || ev.type === "agent_settled") return;
      }
    } catch {
      // best effort: the caller raises the timeout either way
    }
  }

  /** The next event, or null when none arrives within timeoutMs. */
  private async nextEvent(timeoutMs: number): Promise<Json | null> {
    const ev = await this.events.get(timeoutMs);
    if (ev === null) return null;
    if (ev.type === PROCESS_EXIT) {
      // Keep the sentinel visible to any later wait on this dead session.
      this.events.put(ev);
      throw new PiRpcError(
        `[${this.agentName}] pi process exited unexpectedly; ` +
          `stderr tail: ${this.stderrTail.join("").slice(-800)}`,
      );
    }
    return ev;
  }

  private readStdout(proc: ChildProcessWithoutNullStreams): void {
    proc.stdout.setEncoding("utf8");
    let buf = "";
    proc.stdout.on("data", (chunk: string) => {
      buf += chunk;
      let idx: number;
      while ((idx = buf.indexOf("\n")) !== -1) {
        this.handleLine(buf.slice(0, idx));
        buf = buf.slice(idx + 1);
      }
    });
    proc.on("close", () => {
      this.ended = true;
      if (buf) this.handleLine(buf);
      for (const w of this.waiters.values()) {
        w.reject(new PiRpcError(`[${this.agentName}] session ended before response`));
      }
      this.waiters.clear();
      this.events.put({ type: PROCESS_EXIT });
    });
  }

  private handleLine(line: string): void {
    const text = line.endsWith("\r") ? line.slice(0, -1) : line;
    if (!text.trim()) return;
    let obj: unknown;
    try {
      obj = JSON.parse(text);
    } catch {
      return;
    }
    if (typeof obj !== "object" || obj === null || Array.isArray(obj)) return;
    const rec = obj as Json;
    if (rec.type === "response" && "id" in rec) {
      const w = this.waiters.get(String(rec.id));
      this.waiters.delete(String(rec.id));
      w?.resolve(rec);
    } else {
      this.events.put(rec);
    }
  }

  private drainStderr(proc: ChildProcessWithoutNullStreams): void {
    proc.stderr.setEncoding("utf8");
    proc.stderr.on("data", (chunk: string) => {
      this.stderrTail.push(chunk);
      if (this.stderrTail.length > 200) this.stderrTail.shift();
    });
  }
}
